app/src/org/commcare/android/database/HybridFileBackedSqlStorage.java

```diff
@@ -0,0 +1,544 @@
+            throws NoSuchElementException, InvalidIndexException {
+        return getRecordForValues(new String[]{rawFieldName}, new Object[]{value});
+    }
+
+    @Override
+    public byte[] readBytes(int id) {
+        Cursor cur = getDbOrThrow().query(table, dataColumns,
+                DatabaseHelper.ID_COL + "=?",
+                new String[]{String.valueOf(id)}, null, null, null);
+
+        InputStream is = null;
+        try {
+            cur.moveToFirst();
+            byte[] serializedObj = cur.getBlob(cur.getColumnIndexOrThrow(DatabaseHelper.DATA_COL));
+            if (serializedObj != null) {
+                return serializedObj;
+            } else {
+                String filename = cur.getString(cur.getColumnIndexOrThrow(DatabaseHelper.FILE_COL));
+                byte[] aesKeyBlob = cur.getBlob(cur.getColumnIndexOrThrow(DatabaseHelper.AES_COL));
+                is = getInputStreamFromFile(filename, aesKeyBlob);
+                if (is == null) {
+                    throw new RuntimeException("Unable to open and decrypt file: " + filename);
+                }
+
+                return StreamsUtil.getStreamAsBytes(is);
+            }
+        } finally {
+            if (is != null) {
+                try {
+                    is.close();
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
+            }
+            if (cur != null) {
+                cur.close();
+            }
+        }
+    }
+
+    @Override
+    public void write(Persistable persistable) {
+        if (persistable.getID() != -1) {
+            update(persistable.getID(), persistable);
+            return;
+        }
+        SQLiteDatabase db = getDbOrThrow();
+
+        try {
+            db.beginTransaction();
+
+            long insertedId;
+            ByteArrayOutputStream bos = writeExternalizableToStream(persistable);
+            try {
+                if (blobFitsInDb(bos)) {
+                    // serialized object small enough to fit in db
+                    ContentValues contentValues = helper.getNonDataContentValues(persistable);
+                    contentValues.put(DatabaseHelper.DATA_COL, bos.toByteArray());
+                    insertedId = db.insertOrThrow(table, DatabaseHelper.DATA_COL, contentValues);
+                } else {
+                    // store serialized object in file and file pointer in db
+                    File dataFile = HybridFileBackedSqlHelpers.newFileForEntry(dbDir);
+                    ContentValues contentValues = helper.getNonDataContentValues(persistable);
+                    contentValues.put(DatabaseHelper.FILE_COL, dataFile.getAbsolutePath());
+                    byte[] key = generateKeyAndAdd(contentValues);
+                    insertedId = db.insertOrThrow(table, DatabaseHelper.FILE_COL, contentValues);
+
+                    writeStreamToFile(bos, dataFile.getAbsolutePath(), key);
```

Comment:
There's no way to undo this op, currently. An atomic structure for this method would be

* Create DB row/record for "Remove file at PATH on next clean"
* Open Transaction
* Write Row
* Delete "Remove file" row
* Close transaction